src/main/java/xbot/common/trajectory/SwerveSimpleTrajectoryLogic.java

```diff
@@ -263,8 +282,102 @@ private List<XbotSwervePoint> getVelocityAdjustedSwervePoints(
         return velocityAdjustedPoints;
     }
 
+    private List<XbotSwervePoint> getKinematicsAdjustedSwervePoints(
+            XbotSwervePoint initialPoint,
+            List<XbotSwervePoint> swervePoints) {
+
+        ArrayList<XbotSwervePoint> adjustedPoints = new ArrayList<>();
+
+        SwerveKinematicsCalculator calculator = null;
+        for (int i = 0; i < swervePoints.size(); i++) {
+            XbotSwervePoint previous = initialPoint;
+            var current = swervePoints.get(i);
+            if (i > 0) {
+                // If we've moved on to later points, we can now safely get previous entries in the list.
+                previous = swervePoints.get(i - 1);
+                // Calculate the initial velocity of current node
+                current.setKinematics(current.kinematics.kinematicsWithNewVi(calculator.getVelocityAtFinish()));
+            }
+            double distance = previous.getTranslation2d().getDistance(current.getTranslation2d());
+            calculator = new SwerveKinematicsCalculator(
+                    0,
+                    distance,
+                    current.getKinematics()
+            );
+            double adjustedDuration = calculator.getTotalOperationTime();
+
+            if (adjustedDuration > 0) {
+                XbotSwervePoint point = new XbotSwervePoint(current.keyPose, adjustedDuration);
+                point.setKinematics(current.getKinematics());
+                adjustedPoints.add(point);
+            }
+        }
+
+        if (adjustedPoints.size() == 0) {
+            var dummyPoint = new XbotSwervePoint(initialPoint.keyPose, 0.05);
+            adjustedPoints.add(dummyPoint);
+        }
+
+        return adjustedPoints;
+    }
+
+    private List<XbotSwervePoint> getGlobalKinematicsAdjustedSwervePoints(
```

Comment:
⭐ from the name at first blush I'm not sure what the global part of this is and how it varies from `getKinematicsAdjustedSwervePoints` without reading the code more deeply. Is there a more description name or a comment that could explain it?